File: C++/2058. Find the Minimum and Maximum Number of Nodes Between Critical Points.cpp
class Solution {
public:
    vector<int> nodesBetweenCriticalPoints(ListNode* head) {       
        vector<int> dummy({-1, -1});
        vector<int> res;         
        ListNode *prev = head;
        ListNode *curr = head->next;
        ListNode *next = curr->next;
        if(next == NULL){
            return dummy;
        }
        int i = 1;
        while(next != NULL){
            i++;
            if((prev->val < curr->val && curr->val > next->val) || (prev->val > curr->val && curr->val < next->val))
                res.push_back(i);
            prev = curr;
            curr = next;
            next = next->next;  
        }
        if(res.size() < 2){
            return dummy;
        }
        int max_val = res.back() - res.front();
        int min_val = 10000000;
        for(int i = 1; i < res.size(); ++i)
            min_val = min(min_val, res[i] - res[i - 1]);
        res.clear();
        res.push_back(min_val);
        res.push_back(max_val);     
        return res;
    }
};

File: JAVA/2058. Find the Minimum and Maximum Number of Nodes Between Critical Points.java
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        List<Integer> list = new ArrayList<>();
        int prev = head.val;
        int index = 1;
        head = head.next;
        while(head != null){
            int curr = head.val;
            if(head.next != null){
                int next = head.next.val;
                if((curr>prev && curr>next) || (curr<prev && curr<next)){
                    list.add(index);
                }
            }
            head = head.next;
            index++;
            prev = curr;
        }
        int n = list.size();
        if(n < 2){
            return new int[] {-1,-1};
        }
        int min = Integer.MAX_VALUE;
        for(int i=0; i<n-1; i++){
            min = Math.min(min,list.get(i+1)-list.get(i));
        }
        int max = list.get(n-1) - list.get(0);
        return new int[] {min,max};
    }
}
